Shader refactoring

bkoropoff opened the pull request:
- Allow using #include in shaders
- Allow passing defines
- Handle uniforms declaratively
- Use shader stack

kraflab commented on prboom2/src/gl_shader.c:
```diff
@@ -53,217 +54,534 @@
-static GLShader *active_shader = NULL;
-
-static GLShader* gld_LoadShader(const char *vpname, const char *fpname);
+static void ShaderSrcInit(shdr_src_t* src)
```
Are you using a different capitalization because it's static? I experimented with this at some point but ultimately decided I prefer the `prefix_FunctionName` style across the board.

bkoropoff replied:
Should these be `gld` or `glsl` functions?  The file already contains both.

kraflab replied:
I guess pick whichever you like. Probably we should change all of these to the same one at some point, can be done separately.

kraflab commented on prboom2/src/gl_shader.c:
```diff
@@ -53,217 +54,534 @@
+  GLint* lens;
+  unsigned int size;
+  unsigned int cap;
+} shdr_src_t;
```
Let's take this opportunity to use deobfuscated names like `shader_source_t`

kraflab commented on prboom2/src/gl_shader.c:
```diff
@@ -53,217 +54,534 @@
 }
 
-static int ReadLump(const char *filename, const char *lumpname, char **buffer)
+static void ShaderSrcProcess(shdr_src_t* src, const GLchar* text, GLint len,
```
Basically, we have our own preprocessor?

bkoropoff replied:
For `#include`, yeah.  Everything else is handled by OpenGL.

kraflab commented on prboom2/src/gl_shader.c:
```diff
@@ -53,217 +54,534 @@
 
-  idx = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "tex");
-  GLEXT_glUniform1iARB(idx, 0);
+static dboolean GetLine(GLchar const** text, GLint* len, GLchar const** line, GLint* llen)
```
I don't think this file should have custom string handling code. We have `dsda_SplitString(text, "\n\r")` to get the text line by line (e.g., in `dsda_LoadWadPreferences`). If something more robust is required, a separate parser file using the scanner might be safer (see `udmf.cpp` for an example).

bkoropoff replied:
`dsda_SplitString` is kind of heavy:

- The entire shader lump has to be copied and NUL-terminated before calling it
- It does two passes over the string
- It allocates and splits in one go instead of iterating over the string incrementally
- It modifies the string in place (via `strtok`), so the `\n` needs to be added back to each line before passing it to `glShaderSource` (another copy for each source line).
- `glShaderSource` accepts strings as pointer + length (no NUL required) and does a final copy and concatenation internally, so this is all wasted

It's fine for conveniently splitting small strings, but not great for operating on a full file.

I don't really want to use a full lexical analyzer on the source because I'm trying to interpret as little GLSL syntax as possible.  I just want to expand `#include`s and insert `#define`s in the right place (these are really the two major oversights in the GLSL preprocessor).

kraflab replied:
Even if the shader was thousands of lines, would it be measurable? :^)

bkoropoff replied:
It would be a bunch of extra bookkeeping in the rest of the code (allocs, copies, cleanup) just to get rid of the `GetLine` function, which is extremely simple.  I think it would be better for the utility library to grow pointer/length string functions instead.  NUL-terminated strings are just not very flexible, e.g. you can't create a substring without mutating or allocating and copying.

kraflab replied:
That would also be fine with me - basically if there are missing utility tasks that aren't specific to the shader then it would be best to create them separately. The caveat would be things that the scanners already do, because I want to avoid having multiple different implementations of the same task.

bkoropoff replied:
The scanner isn't a good match because it's both overkill (I don't need to parse the vast majority of token types, because I'm doing very light preprocessing of the source), but doesn't handle things I do need (C preprocessor-style directives).  I'll see about adding the start of a pointer/length API to the DSDA utilities.

bkoropoff replied:
I added the start of a string view API, it's fairly clean and complements the `dsda_string` API, sort of like `string`/`string_view` in C++.

bkoropoff commented:
Added some additional refactoring so that a single specification of uniforms is used for both location resolution and value setting/tracking.

kraflab commented on prboom2/data/lumps/gls_main.lmp:
```diff
@@ -1,3 +1,15 @@
+#version 110
+
+#ifdef VERTEX
```
Why do we want two files squeezed into one like this?

bkoropoff replied:
Putting all shader stages in the same file is typical in HLSL, [Slang](https://github.com/shader-slang/slang/blob/master/examples/triangle/shaders.slang), [WGSL](https://github.com/gfx-rs/wgpu-native/blob/trunk/examples/triangle/shader.wgsl), [Metal](https://metalbyexample.com/up-and-running-2/), etc.  Since the stages are intimately linked, you can keep everything in one place, specify `varying`s only once, etc.  It's like not having to have separate `.c`/`.h` files.  The vertex stage we use right now is extremely trivial, so it's not a huge benefit, but if we ever start passing additional data through to the fragment stage for different shaders, having the code in one place is tidier.

kraflab replied:
Understood 👍

bkoropoff committed:
GL: fix init order

bkoropoff committed:
dsda:utility: add start of string view API

String views are immutable views into existing strings.
Because they use a pointer/length representation and don't allocate,
taking substrings is "free".

bkoropoff committed:
GL: refactor shader handling

- Allow using #include in shaders
- Allow passing defines
- Handle uniforms declaratively
- Use shader stack

bkoropoff commented:
Let me know if you find the `strview` API a good direction

kraflab commented:
Hopefully I'll find some time this weekend to go through everything.

kraflab committed:
Remove trailing whitespace

kraflab committed:
Refactor utility string view

kraflab commented on prboom2/src/dsda/utility/string_view.c:
```diff
@@ -0,0 +1,125 @@
+  return false;
+}
+
+dboolean dsda_AdvanceStringViewToNextLine(dsda_string_view_t* sv, dsda_string_view_t* line) {
```
What do you think about switching this to something like `dsda_GetStringViewLine` that sets line to the _current_ line and advances the string view forward (so treating it more like reading from a stream)?

bkoropoff replied:
That's what this function does.  `line` is set to the the line starting at the current position of `sv`, and `sv` is advanced past it.

kraflab replied:
Ah ya, now I see the line is in the before slot. I think the comment before wasn't clear. Maybe the name should indeed be `dsda_GetStringViewLine` then.

bkoropoff replied:
Having `Next` in the name emphasizes the iterator-like behavior, but yeah.

kraflab commented on prboom2/src/dsda/utility/string_view.h:
```diff
@@ -0,0 +1,58 @@
+
+dboolean dsda_StringViewStartsWithPrefix(const dsda_string_view_t* sv, const char* prefix);
+
+void dsda_StringViewAfterChars(const dsda_string_view_t* sv, const char* chars,
```
The comment for this said it removed chars from sv, but it doesn't appear to do that - it just sets the after to after the chars. Is it supposed to?

bkoropoff replied:
It should "return" a new string, since there's no reason to require that `sv` be modified in place, and you can just use it as the output if that's what you want.

This function should really be called "TrimLeft" or "StripBefore" or something, as it's intended to work like [`lstrip`](https://docs.python.org/3/library/stdtypes.html#str.lstrip) in python.  There would be a corresponding function to strip from the right side of the string, and one that does both at once.  It's usually used to discard leading/trailing whitespace.

kraflab replied:
It would confuse me to use that name and have it not mutate or return newly allocated data, that's why I've changed the names like this.

bkoropoff replied:
What would you call `rstrip`, which strips from the right side of the string, or `strip`, which does both?

kraflab replied:
dsda_StringViewBeforeChars I guess :^)
And dsda_StringViewBetweenChars 😸

kraflab replied:
By the way, something for you to keep in mind in the future. If you are going to do something that you can encapsulate and decouple cleanly, where you would benefit from the c++ stuff, it's completely fine with me if you do it in c++. We have this for the udmf and umapinfo parsers, and preprocessing text is probably another thing that can be cleanly isolated.

kraflab commented on prboom2/src/dsda/utility/string_view.h:
```diff
@@ -0,0 +1,58 @@
+void dsda_StringViewAtOffset(const dsda_string_view_t* sv, size_t offset, dsda_string_view_t* ofs);
+
+// Splits `sv` at the first instance of `c`, putting everything up to and including it in `before`
+// and everything after in `after`.  `before` or `after` may safely alias `sv`.  If `c` does not
```
I'm confused by the use of alias in this file. What does it mean in this context?

bkoropoff replied:
That's C jargon for "point to the same object in memory".

kraflab replied:
Ah now I see what you mean. In other words you can safely pass the source as the before / after as well.

kraflab commented on prboom2/src/dsda/utility/string_view.c:
```diff
@@ -0,0 +1,125 @@
+  const char* cur;
+  size_t size;
+
+  for (cur = sv->string, size = sv->size; size && strchr(chars, *cur); ++cur, --size);
```
Should we have the "split" functions take a group of chars as well and consolidate everything?

bkoropoff replied:
The split by char functions could take a group of chars, but it wouldn't obviate the need for this function, since split always stops at the first instance (on purpose, since you want to preserve line counts, etc.), whereas this eats through consecutive instances.

kraflab replied:
Do we need that behaviour or should we discard groups of chars anyway?

bkoropoff replied:
Yeah, the one-at-a-time behavior is more general (you can always skip trivial fields if you don't care about them), and what we want in this case (to avoid deleting blank lines from the source).

bkoropoff commented on prboom2/src/dsda/utility/string_view.h:
```diff
@@ -0,0 +1,58 @@
+// is emptied.  Returns `false` if there was no more data, i.e. `sv` was empty.
+dboolean dsda_AdvanceStringViewToNextLine(dsda_string_view_t* sv, dsda_string_view_t* line);
+
+dboolean dsda_StringViewStartsWithPrefix(const dsda_string_view_t* sv, const char* prefix);
```
The original name `StartsWithCStr` is intentional, as you'd also want a variant that checks for a `string_view` as a prefix.

kraflab replied:
I'd probably call that variant `StartsWithViewPrefix` or something 🤔

bkoropoff replied:
`Prefix` is redundant at any rate.

kraflab replied:
Good point 😄

kraflab committed:
Update string view naming

kraflab commented on prboom2/src/SDL/i_video.c:
```diff
@@ -1412,6 +1409,9 @@ void I_UpdateVideoMode(void)
     gld_Init(SCREENWIDTH, SCREENHEIGHT);
   }
 
+  ST_SetResolution();
```
Why did this move? Was there a bug?

bkoropoff replied:
Yeah, these can end up making drawing calls before GL is initiated.  In my lightmode branch it caused attempts to use shaders before they are allocated/loaded.

kraflab commented on prboom2/src/gl_main.c:
```diff
@@ -823,14 +824,15 @@ void gld_DrawWeapon(int weaponlump, vissprite_t *vis, int lightlevel)
   }
   glColor3f(1.0f,1.0f,1.0f);
-  gld_EndFuzz();
+  if (/*(viewplayer->mo->flags & MF_SHADOW) && */!vis->colormap)
```
Bug fix?

bkoropoff replied:
Without this check to mirror the one for `StartFuzz`, you can get unbalanced push/pop of the shader stack.

kraflab commented:
I'm getting dangerously close to removing msvc from the CI 😅

kraflab committed:
Use NullShader

kraflab commented:
Looks good to me. Any other concerns or should we merge it? I would probably prefer the preprocessor to be in its own file, but I'll take care of that separately, not sure yet if I want to do it.

kraflab commented on prboom2/src/gl_shader.c:
```diff
@@ -53,217 +54,551 @@
 }
 
-static int ReadLump(const char *filename, const char *lumpname, char **buffer)
+static void glsl_ShaderSrcProcess(shader_source_t* src, const GLchar* text,
```
It occurs to me that we could run the preprocessor as part of the build step when constructing the wad so the lump is already processed, and there's nothing to do at run time. I would leave the current PR as it is and take that separately if we decide to do it. What do you think?

bkoropoff replied:
If we invest in build-time shader processing, we'd want to run glslangValidator on it to validate the shaders, in which case we could use it to do the preprocessing (`glslangValidator -E`).  This would bloat the data going into `dsda-doom.wad`, if that matters.

kraflab commented:
Thanks!
